misoc/cores/icap.py

```diff
@@ -0,0 +1,185 @@
+from migen import *
+from migen.genlib.misc import timeline
+
+from misoc.interconnect import wishbone
+from misoc.interconnect.csr import AutoCSR, CSRStorage, CSRStatus
+
+IDCODE = 0x03631093 # 7a100t
+ICAP_WIDTH = "X32"
+FILE_NAME = "NONE"
+iprog_command_seq_i = [
+            0xFFFFFFFF, # Dummy Word
+            0x5599AA66, # Sync Word
+            0x04000000, # Type 1 NO OP 
+            0x0C400080, # Write WBSTAR 
+            0x00000000, # WBSTAR 
+            0x0C000180, # Write CMD 
+            0x000000F0, # Write IPROG 
+            0x04000000, # Type 1 NO OP  
+        ]
+
+class ICAP(Module, AutoCSR):
+    def __init__(self):
+        self.trigger = CSRStorage()
+
+        ###
+
+        O = Signal(32)      # 32-bit output: Configuration data output bus
+        CLK = Signal()      # 1-bit input: Clock Input
+        CSIB = Signal()     # 1-bit input: Active-Low ICAP Enable
+        I = Signal(32)      # 32-bit input: Configuration data input bus
+        RDWRB = Signal()    # 1-bit input: Read/Write Select input
+        state = Signal(4)   # state machine register
+
+        # clock generation
+        div = 4
+        counter = Signal(max=div)
+        self.sync += [
+            If(counter == div//2 - 1,
+                CLK.eq(1)
+            ),
+            If(counter == div - 1,
+                counter.eq(0),
+                CLK.eq(0)
+            ).Else(
+                counter.eq(counter + 1),
+            ),
+        ]
+
+        self.clock_domains.icap = ClockDomain(reset_less=True)
+        self.comb += ClockSignal("icap").eq(CLK)
+
+        # rising edge detection
```

Comment:
Needs a synchronizer. See the ``PulseSynchronizer`` module in Migen and e.g. https://www.eetimes.com/understanding-clock-domain-crossing-issues/